python/ray/llm/_internal/serve/serving_patterns/data_parallel/dp_group_manager.py

```diff
@@ -0,0 +1,313 @@
+
+        Args:
+            replica_rank: The replica's rank info from serve.get_replica_context().rank.
+            replica_id: Unique identifier for the replica.
+
+        Returns:
+            Tuple of (dp_rank, group_index).
+        """
+        dp_rank = self._get_dp_rank(
+            replica_rank, self.dp_group_size, self.dp_size_per_node
+        )
+        group_index = self._get_group_index(
+            replica_rank, self.dp_group_size, self.dp_size_per_node
+        )
+
+        async with self._group_info_lock:
+            if group_index not in self._group_info:
+                self._group_info[group_index] = GroupInfo()
+
+            group = self._group_info[group_index]
+
+            # Get replicas to kill if double-registration is detected
+            replicas_to_kill = self._handle_double_registration(
+                group, group_index, dp_rank, replica_id
+            )
+
+            # Register the new replica
+            group.dp_rank_to_replica_id[dp_rank] = replica_id
+
+            logger.info(
+                f"Registered replica {replica_id} to group {group_index} "
+                f"with dp_rank {dp_rank} (replica_rank={replica_rank})"
+            )
+
+        # Kill replicas outside the lock to avoid blocking
+        if replicas_to_kill:
+            await self._kill_replicas(replicas_to_kill, group_index)
+
+        return dp_rank, group_index
+
+    def _reset_group(self, group: GroupInfo) -> None:
+        """Reset a group's state for a fresh restart.
+
+        Clears the dp_rank_to_replica_id mapping and resets the master info
+        so a new master can be elected.
+
+        Must be called while holding _group_info_lock.
+        """
+        # If there are any waiters, wake them so they can re-check state.
+        group.master_info_event.set()
+        group.master_info_event.clear()
+
+        group.dp_rank_to_replica_id.clear()
+        group.master_info = None
+
+    def _handle_double_registration(
+        self,
+        group: GroupInfo,
+        group_index: int,
+        dp_rank: int,
+        replica_id: str,
+    ) -> List[str]:
+        """Check for and handle double-registration.
+
+        If a new replica is trying to register for a DP rank already occupied
+        by a different replica, this indicates a replica failure/restart.
+        Collects replicas to kill and resets the group for a clean restart.
+
+        Must be called while holding _group_info_lock.
+
+        Args:
+            group: The group info object.
+            group_index: The index of the group.
+            dp_rank: The data parallel rank being registered.
+            replica_id: The replica ID attempting to register.
+
+        Returns:
+            List of replica IDs to kill (empty if no double-registration).
+        """
+        existing_replica_id = group.dp_rank_to_replica_id.get(dp_rank)
+        if existing_replica_id is None or existing_replica_id == replica_id:
```

Comment:
should we not count as double registeration if the returned replica id is the same as the current replica_id? I think this could be also a failure mode, since this should never actually happen. right?